xayn-ai/src/embedding/utils.rs

```diff
@@ -1,90 +1,166 @@
-use std::ops::Deref;
-
-use ndarray::{ArrayBase, Data, Ix1};
+use ndarray::{Array1, Array2, ArrayBase, Data, Ix1};
 use rubert::Embedding1;
 
 pub(crate) type Embedding = Embedding1;
 
-/// Computes the l2 norm (euclidean metric) of the difference of two vectors.
+/// Computes the l2 norm (euclidean metric) of a vector.
 ///
 /// # Panics
-/// Panics if the vectors didn't consist of all real values.
-pub fn l2_distance<A, B, S>(a: &A, b: &B) -> f32
+/// Panics if the vector doesn't consist solely of real values.
+pub fn l2_norm<S>(a: ArrayBase<S, Ix1>) -> f32
 where
-    A: Deref<Target = ArrayBase<S, Ix1>>,
-    B: Deref<Target = ArrayBase<S, Ix1>>,
     S: Data<Elem = f32>,
 {
-    let difference = a.deref() - b.deref();
-    let distance = difference.dot(&difference).sqrt();
-
-    if distance.is_nan() || distance.is_infinite() {
-        panic!(
-            "vectors must consist of real values only, but got\na: {:?}\nb: {:?}",
-            a.deref(),
-            b.deref(),
-        );
-    }
+    let norm = a.dot(&a).sqrt();
+    assert!(
+        norm.is_finite(),
+        "vector must consist of real values only, but got:\n{:?}",
+        a,
+    );
+
+    norm
+}
 
-    distance
+/// Computes the l2 norm (euclidean metric) of the difference of two vectors.
+///
+/// # Panics
+/// Panics if the vectors don't consist solely of real values or their shapes don't match.
+pub fn l2_distance<S>(a: ArrayBase<S, Ix1>, b: ArrayBase<S, Ix1>) -> f32
+where
+    S: Data<Elem = f32>,
+{
+    l2_norm(&a - &b)
 }
 
 /// Computes the arithmetic mean of two vectors.
 ///
 /// # Panics
-/// Panics if the vectors do not consist solely of real values.
-pub fn mean<A, S>(a: &A, b: &A) -> Embedding
+/// Panics if the vectors don't consist solely of real values or their shapes don't match.
+pub fn mean<S>(a: ArrayBase<S, Ix1>, b: ArrayBase<S, Ix1>) -> Array1<f32>
 where
-    A: Deref<Target = ArrayBase<S, Ix1>>,
     S: Data<Elem = f32>,
 {
-    let mean = 0.5 * (a.deref() + b.deref());
-    if mean.iter().any(|elt| elt.is_nan() || elt.is_infinite()) {
-        panic!(
-            "vectors must consist of real values only, but got\na: {:?}\nb: {:?}",
-            a.deref(),
-            b.deref(),
-        );
+    let mean = 0.5 * (&a + &b);
+    assert!(
+        mean.iter().copied().all(f32::is_finite),
+        "vectors must consist of real values only, but got\na: {:?}\nb: {:?}",
+        a,
+        b,
+    );
+
+    mean
+}
+
+/// Computes the pairwise cosine similarities of vectors.
+///
+/// # Panics
+/// Panics if the vectors don't consist solely of real values or their shapes don't match.
+#[allow(dead_code)]
+pub fn pairwise_cosine_similarity<S>(a: &[ArrayBase<S, Ix1>]) -> Array2<f32>
+where
+    S: Data<Elem = f32>,
+{
+    let norms = a.iter().map(|a| l2_norm(a.view())).collect::<Vec<_>>();
+    let size = a.len();
+    let mut similarities = Array2::ones((size, size));
```

Comment:
Just to be sure, the  problems with QEMU  only applies to `::zeros()`?

Reply:
correct, only the `zeros` will lead to the `memset` path, all other values will lead to the `memcpy` path.